service/history/transferQueueTaskExecutorBase_test.go

```diff
@@ -0,0 +1,216 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"context"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/service/history/tests"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func TestOnlyDeleteOpenExecutions(t *testing.T) {
+	testCases := []struct {
+		Name                    string
+		EnsureCloseBeforeDelete bool
+		CloseTransferTaskIdSet  bool
+		InActiveCluster         bool
+		MultiCursorQueue        bool
+		CloseTaskIsAcked        bool
+		ShouldDelete            bool
+	}{
+		{
+			Name:                    "skip the check",
+			EnsureCloseBeforeDelete: false,
+			ShouldDelete:            true,
+		},
+		{
+			Name:                    "no task id",
+			EnsureCloseBeforeDelete: true,
+			CloseTransferTaskIdSet:  false,
+			ShouldDelete:            true,
+		},
+		{
+			Name:                    "passive cluster",
+			EnsureCloseBeforeDelete: true,
+			CloseTransferTaskIdSet:  true,
+			InActiveCluster:         false,
+			ShouldDelete:            true,
+		},
+		{
+			Name:                    "single cursor queue unacked",
+			EnsureCloseBeforeDelete: true,
+			CloseTransferTaskIdSet:  true,
+			InActiveCluster:         true,
+			MultiCursorQueue:        false,
+			CloseTaskIsAcked:        false,
+			ShouldDelete:            false,
+		},
+		{
+			Name:                    "single cursor queue acked",
+			EnsureCloseBeforeDelete: true,
+			CloseTransferTaskIdSet:  true,
+			InActiveCluster:         true,
+			MultiCursorQueue:        false,
+			CloseTaskIsAcked:        true,
+			ShouldDelete:            true,
+		},
+		{
+			Name:                    "multicursor queue unacked",
+			EnsureCloseBeforeDelete: true,
+			CloseTransferTaskIdSet:  true,
+			InActiveCluster:         true,
+			MultiCursorQueue:        true,
+			CloseTaskIsAcked:        false,
+			ShouldDelete:            false,
+		},
+		{
+			Name:                    "multicursor queue acked",
+			EnsureCloseBeforeDelete: true,
+			CloseTransferTaskIdSet:  true,
+			InActiveCluster:         true,
+			MultiCursorQueue:        true,
+			CloseTaskIsAcked:        true,
+			ShouldDelete:            true,
+		},
+	}
+	for _, c := range testCases {
+		t.Run(c.Name, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
```

Comment:
> // New in go1.14+, if you are passing a *testing.T into NewController function you no
// longer need to call ctrl.Finish() in your test methods.

Reply:
Oh sweet, that's nice to know!